### PR TITLE
createFunction() and createProcedure() (MLDB-1192)

### PR DESCRIPTION
**Pull request by jeremybarnes**
Improves the usability of the JS bindings and includes some refactorings under the hood.  Functionality exposed on the Procedure and Function JS objects is currently sparse, but we have a foundation to build on now.

### CODE REVIEW

**Comment by nicolaskruchten** [comment]
could we get at least one small test?

File: container_files/assets/www/doc/builtin/lang/Javascript.md [hunk 1]
```diff
@@ -243,4 +258,33 @@ This object has the following methods defined:
...
+- `procedure.type()` returns the type of the procedure
+- `procedure.config()` returns the configuration of the procedure
+
+### Function objects
+
+A function object is created similarly to procedure and dataset objects.
+
+It has the following methods defined:
+
+- `function.run(args)` runs the function, with the given arguments,
```
Comment (guyd) [inline container_files/assets/www/doc/builtin/lang/Javascript.md]: Why not use apply here for consistency with the REST API?

File: plugins/lang/js/function_js.cc [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+registerMe()
+{
+    using namespace v8;
+
+    HandleScope scope;
+
+    auto fntmpl = CreateFunctionTemplate("Function");
+    auto prototmpl = fntmpl->PrototypeTemplate();
+
+    prototmpl->Set(String::New("status"), FunctionTemplate::New(status));
```
Comment (guyd) [inline plugins/lang/js/function_js.cc]: This function was not documented above.

File: plugins/lang/js/procedure_js.cc [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+registerMe()
+{
+    using namespace v8;
+
+    HandleScope scope;
+
+    auto fntmpl = CreateFunctionTemplate("Procedure");
+    auto prototmpl = fntmpl->PrototypeTemplate();
+
+    prototmpl->Set(String::New("status"), FunctionTemplate::New(status));
```
Comment (guyd) [inline plugins/lang/js/procedure_js.cc]: This function is also not covered in the documentation.  Is that intentional?

File: server/server.mk [hunk 4]
```diff
@@ -25,7 +25,7 @@ LIBMLDB_SOURCES:= \
 	bound_queries.cc \
 	script_output.cc \
 	forwarded_dataset.cc \
-	serial_function.cc
+	serial_function.cc \
```
Comment (guyd) [inline server/server.mk]: is the trailing slash required?

File: sql/sql_utils.cc [hunk 5]
```diff
@@ -0,0 +1,43 @@
+/** sql_utils.cc
+    Jeremy Barnes, 30 November 2015
+    Copyright (c) 2015 Datacratic Inc.  All rights reserved.
+
```
Comment (guyd) [inline sql/sql_utils.cc]: This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.

File: sql/sql_utils.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+/** sql_utils.h                                                    -*- C++ -*-
+    Jeremy Barnes, 30 November 2015
+    Copyright (c) 2015 Datacratic Inc.  All rights reserved.
+
```
Comment (guyd) [inline sql/sql_utils.h]: This file is part of MLDB. Copyright 2015 Datacratic. All rights reserved.

File: container_files/assets/www/doc/builtin/lang/Javascript.md [hunk 7]
```diff
@@ -157,6 +157,21 @@ apply a JSON patch:
   object.  The return value is an array; element 0 is the patched object,
   and element 1 is the set of diffs that couldn't be applied.
 
+### SQL
+
+The following functions aid in working with SQL from within Javascript.
+
+- `mldb.sql(<sql statement>)` will parse and run the given query (which is
```
Comment (guyd) [inline container_files/assets/www/doc/builtin/lang/Javascript.md]: Consider naming it _query_ for consistency with the REST API.

**Comment by guyd** [comment]
I also vote for one small test!

File: server/server.mk (hunk 4 above)
Comment (jeremybarnes) [inline server/server.mk]: It reduces noise on adding a new source file as it's a one line diff.  But no, not required.

File: container_files/assets/www/doc/builtin/lang/Javascript.md (hunk 1 above)
Comment (jeremybarnes) [inline container_files/assets/www/doc/builtin/lang/Javascript.md]: It was wrong; the current word is `call`.  Apply comes from when we had the concept of "block"s, we should probably rename it as call everywhere due to us having functions now.

File: plugins/lang/js/function_js.cc (hunk 2 above)
Comment (jeremybarnes) [inline plugins/lang/js/function_js.cc]: Added doc to all of the entity types.

File: plugins/lang/js/procedure_js.cc (hunk 3 above)
Comment (jeremybarnes) [inline plugins/lang/js/procedure_js.cc]: Added doc.

File: sql/sql_utils.cc (hunk 5 above)
Comment (jeremybarnes) [inline sql/sql_utils.cc]: Fixed.

File: sql/sql_utils.h (hunk 6 above)
Comment (jeremybarnes) [inline sql/sql_utils.h]: Fixed.

File: container_files/assets/www/doc/builtin/lang/Javascript.md (hunk 7 above)
Comment (jeremybarnes) [inline container_files/assets/www/doc/builtin/lang/Javascript.md]: Modified.

**Comment by jeremybarnes** [comment]
Added a small test and fixed up things identified in the previous comments.

**Comment by jeremybarnes** [comment]
Test failure is the Pytanic test which will fix itself when merged into master.

**Comment by guyd** [comment]
+1

**Pull request merged**